### ISSUE TITLE
[SLPVectorizer] Miscompile due to incorrect nuw handling

### ISSUE THREAD

**Issue by max-quazan** [llvm/llvm-project#55688]
Godbolt repro: https://godbolt.org/z/qeeo1P5Ej
Test:
`opt -passes=slp-vectorizer -S`

```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target triple = "x86_64-unknown-linux-gnu"

define i32 @test() {
entry:
  br label %loop

loop:
  %tmp = phi i32 [ -2, %entry ], [ %tmp13, %loop ]
  %tmp2 = add nuw nsw i32 %tmp, 1
  %tmp3 = add nsw i32 %tmp2, 1
  %tmp4 = add nuw nsw i32 %tmp3, 1
  %tmp5 = add nuw nsw i32 %tmp4, 1
  %tmp6 = add nuw nsw i32 %tmp5, 1
  %tmp7 = add nuw nsw i32 %tmp6, 1
  %tmp8 = add nuw nsw i32 %tmp7, 1
  %tmp9 = add nuw nsw i32 %tmp8, 1
  %tmp10 = add nuw nsw i32 %tmp9, 1
  %tmp11 = add nuw nsw i32 %tmp10, 1
  %tmp12 = add nuw nsw i32 %tmp11, 1
  %tmp13 = add nuw nsw i32 %tmp12, 1
  %tmp14 = icmp ugt i32 %tmp12, 12668
  br i1 %tmp14, label %exit, label %loop

exit:
  ret i32 %tmp12
}
```

Output:

```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target triple = "x86_64-unknown-linux-gnu"

define i32 @test() {
entry:
  br label %loop

loop:                                             ; preds = %loop, %entry
  %tmp = phi i32 [ -2, %entry ], [ %tmp13, %loop ]
  %0 = call i32 @llvm.vector.reduce.add.v8i32(<8 x i32> <i32 1, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1, i32 1>)
  %op.rdx = add nuw nsw i32 2, %tmp
  %op.rdx1 = add nuw nsw i32 %0, %op.rdx
  %tmp12 = add nuw nsw i32 %op.rdx1, 1
  %tmp13 = add nuw nsw i32 %tmp12, 1
  %tmp14 = icmp ugt i32 %tmp12, 12668
  br i1 %tmp14, label %exit, label %loop

exit:                                             ; preds = %loop
  ret i32 %tmp12
}

declare i32 @llvm.vector.reduce.add.v8i32(<8 x i32>) #0

attributes #0 = { nocallback nofree nosync nounwind readnone willreturn }
```

Before SLP vectorizer, nuw flag wasn't present in %tmp3:

```llvm
  %tmp = phi i32 [ -2, %bb ], [ %tmp13, %bb1 ]
  %tmp2 = add nuw nsw i32 %tmp, 1
  %tmp3 = add nsw i32 %tmp2, 1
```
because at this moment this value crosses border of zero, and therefore makes unsigned wrap. The whole program is well-defined and exits once the counter exceeds 12668.

After slp:

```llvm
  %op.rdx = add nuw nsw i32 2, %tmp
```
is poison starting 1st iteration when `%tmp = -2` and whole program is UB.

Our triage points at this patch as guilty:

```
commit 7ea03f0b4e4ec5d91d48ba2976f5adc299089ffd
Author: Alexey Bataev <a.bataev@outlook.com>
Date: Thu Nov 18 08:08:01 2021 -0800

[SLP]Improve reductions analysis and emission, part 1.

Currently SLP vectorizer walks through the instructions and selects
3 main classes of values: 1) reduction operations - instructions with same
reduction opcode (add, mul, min/max, etc.), which build the reduction,
2) reduced values - instructions with the same opcodes, but different
from the reduction opcode, 3) extra arguments - all other values,
instructions from the different basic block rather than the root node,
instructions with to many/less uses.

This scheme is not very efficient. It excludes some instructions and all
non-instruction values from the reductions (constants, proficient
gathers), to many possibly reduced values are marked as extra arguments.
Patch improves this process by introducing a bit extended analysis
stage. During this stage, we still try to select 3 classes of the
values: 1) reduction operations - same as before, 2) possibly reduced
values - all instructions from the current block/non-instructions, which
may build a vectorization tree, 3) extra arguments - instructions from
the different basic blocks. Additionally, an extra sorting of the
possibly reduced values occurs to build the scalar sequences which
highly likely will bed vectorized, e.g. loads are grouped by the
distance between them, constants are grouped together, cmp instructions
are sorted by their compare types and predicates, extractelement
instructions are sorted by the vector operand, etc. Also, these groups
are reordered by their length so the longest group is the first in the
list of the possibly reduced values.

The vectorization process tries to emit the reductions for all these
groups. These reductions, remaining non-vectorized possible reduced
values and extra arguments are then combined into the final expression
just like it was before.

Differential Revision: https://reviews.llvm.org/D114171
```

**Comment by max-quazan** [comment]
@alexey-bataev

**Comment by nunoplopes** [comment]
Alive2 agrees that this is a miscompilation: https://alive2.llvm.org/ce/z/v2Fk8b

**Comment by max-quazan** [comment]
Reason is that we call `propagateIRFlags` from `createOp` with arguments: dest = `%op.rdx.second = add i32 2, %tmp`, VL = `{%tmp2 = add nuw nsw i32 %tmp, 1}`. I don't know what the intention was, but it is just wrong.

**Comment by alexey-bataev** [comment]
Yep, correct. Will prepare a fix for it later today.

**Comment by max-quazan** [comment]
Thanks!

**Comment by alexey-bataev** [comment]
Fixed in 10f41a214767219ef369b217f685eb545b80e386
